Codec/Archive/Tar/Index/Utils.hs

```diff
@@ -0,0 +1,41 @@
+module Codec.Archive.Tar.Index.Utils where
+
+import Data.ByteString as BS
+import Control.Exception (assert)
+
+import Data.ByteString.Internal (ByteString(..), unsafeWithForeignPtr, accursedUnutterablePerformIO)
+import GHC.Int (Int(..), Int32)
+import GHC.Word (Word32(..), byteSwap32)
+import Foreign.Storable (peek)
+import GHC.Ptr (castPtr, plusPtr, Ptr)
+import Data.Array.Base
+import Data.Array.Base
+import Data.Array.IO.Internals (unsafeFreezeIOUArray)
+
+-- | Construct a `UArray Word32 Word32` from a ByteString of 32bit big endian
+-- words.
+beToLe :: Word32
+       -- ^ The total array length (the number of 32bit words in the array)
+       -> BS.ByteString
+       -- ^ The bytestring from which the UArray is constructed.
+       -- The content must start in the first byte! (i.e. the meta-data words
+       -- that shouldn't be part of the array must have been dropped already)
+       -> UArray Word32 Word32
+beToLe lenArr (BS fptr _) = accursedUnutterablePerformIO $
```

Comment:
I think it would involve less `unsafe...` stuff if you manually create an appropriate `ByteArray#` from `ByteString` and wrap it into `UArray`.

Reply:
This was our first approach, but then we realised we were just re-inventing `newGenArray` with the unboxed primitives. The amount of "unsafety" was roughly the same, but the code was much, much more complex and some 2 or 3 times longer.

Even more, it ended up being that `newGenArray` was marginally faster than doing the manual construction in our experiments.

Reply:
I'd like to take a look, if you have a branch somewhere. My impression is that all we need is a loop to copy `Word32` from one buffer to another (with bytes swapped).

Reply:
The final version slightly differs (expects the first byte in the array to be the relevant byte, rather than hardcoding the offset), but here is: https://gist.github.com/mpickering/08d322b64d1b88751e762533a2009d21

Reply:
I think it's not too scary, especially taking into account complications with `newGenArray` non-existend before `array-0.5.6.0`.

Reply:
I prefer an implementation which uses the interface from `array` if possible because it would lead to very hard issues to debug if internal representation of UArray changed. It seems that using the higher level combinator to build a mutable array is the right way to use the library if possible.

If you are hard advocating for it to be changed then please say so and the PR can be changed to use the low level implementation.  That's no problem at all.

Perhaps another solution is if there are problems with `array-0.5.6.0` then the implementation of `newGenArray` can be vendored into `tar` so that it works with older `array` versions